torch/csrc/jit/passes/linear_to_matmul_add.cpp

```diff
@@ -0,0 +1,56 @@
+#include <algorithm>
+
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/passes/linear_to_matmul_add.h>
+#include <torch/csrc/jit/passes/tensorexpr_fuser.h>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+void DecomposeLinearToMatmulAdd(Block *b) {
+  // TODO only do the expansion for GPU since CPU fusion isn't on by default
+  // (and thus may slow down instead)
+  std::vector<Node*> linear_nodes;
+  for (auto it = b->nodes().begin(); it != b->nodes().end(); it++) {
+    for (auto* child_block : it->blocks()) {
+      DecomposeLinearToMatmulAdd(child_block);
+    }
+    if (it->kind() == aten::linear) {
+      linear_nodes.push_back(*it);
+    }
+  }
+
+  for (Node* node : linear_nodes) {
+    // If all its uses are supported for fusion, decompose into matmul + add
+    auto uses = node->output()->uses();
+    if (std::all_of(uses.begin(), uses.end(), [](const Use& u) {
+          return tensorexpr::isSupported(u.user);
+        })) {
+      auto g = b->owningGraph();
+      auto matmul_n =
```

Comment:
`linear` actually computes `x A^t + b`, so we also have to insert a transpose node on weight value before using it in the matmul.